aten/src/ATen/native/sparse/cuda/SparseCUDAApplyUtils.cuh

```diff
@@ -0,0 +1,324 @@
+#pragma once
+
+#include <ATen/cuda/detail/TensorInfo.cuh>
+#include <THC/THCNumerics.cuh>
+
+namespace at { namespace native {
+
+namespace apply {
+
+using at::cuda::detail::TensorInfo;
+using indexT = int64_t;
+
+const int WARP_SIZE = 32;
+
+template <typename IndexType, typename Real, typename Op>
+__device__ void applyOp2(
+    Op op, IndexType blockSize,
+    TensorInfo<Real, IndexType> values1, IndexType idx1,
+    TensorInfo<Real, IndexType> values2, IndexType idx2) {
+  for (IndexType k = blockIdx.x * blockDim.x + threadIdx.x;
+       k < blockSize;
+       k += gridDim.x * blockDim.x) {
+    op(values1.data + idx1 * blockSize + k, values2.data + idx2 * blockSize + k);
+  }
+}
+
+template <typename IndexType, typename Real, typename Op>
+__device__ void applyOp3(
+    Op op, IndexType blockSize,
+    TensorInfo<Real, IndexType> values1, IndexType idx1,
+    TensorInfo<Real, IndexType> values2, IndexType idx2,
+    TensorInfo<Real, IndexType> values3, IndexType idx3) {
+  for (IndexType k = blockIdx.x * blockDim.x + threadIdx.x;
+       k < blockSize;
+       k += gridDim.x * blockDim.x) {
+    op(values1.data + idx1 * blockSize + k,
+       values2.data + idx2 * blockSize + k,
+       values3.data + idx3 * blockSize + k);
+  }
+}
+
+template <typename Op, typename IndexType, typename Real>
+__global__ void sparseElementwiseKernel(
+    Op op,
+    TensorInfo<Real, IndexType> dense,
+    TensorInfo<indexT, IndexType> indices,
+    TensorInfo<Real, IndexType> values,
+    const IndexType nnz) {
+  IndexType indskip = indices.strides[0];
+  IndexType valueSize = values.strides[0];
+  for (IndexType linearId = blockIdx.x;
+       linearId < nnz;
+       linearId += gridDim.x) {
+    IndexType index = 0;
+    for (IndexType d = 0; d < indices.sizes[0]; d++) {
+      index = dense.sizes[d] * index + indices.data[d * indskip + linearId];
+    }
+    Real *dst = dense.data + index * valueSize;
+    Real *src = values.data + linearId * valueSize;
+    for (IndexType linearId2 = threadIdx.x; linearId2 < valueSize; linearId2 += blockDim.x) {
+      op(dst + linearId2, src + linearId2);
+    }
+  }
+}
+
+template <typename Op, typename IndexType, typename Real>
+__global__ void sparseElementwiseKernelScalar(
+    Op op,
+    TensorInfo<Real, IndexType> dense,
+    TensorInfo<indexT, IndexType> indices,
+    TensorInfo<Real, IndexType> values,
+    const IndexType nnz) {
+  IndexType indskip = indices.strides[0];
```

Comment:
we decided scalar values/indices aren't supported, right?

Reply:
Yep. What makes you wonder about it?

Reply:
Because you are accessing strides[0] without checking the dimensionality :).

Reply:
I humbly request we defer zero-dim-fix-ifying the functions in this file to a future PR. I literally did not read any of them when porting; they all ported without changes XD

Reply:
but you can't do that, because the tensor representation is different.  In any case, my understanding was this is safe because none of them will be scalars anyway -- is that not correct?

Reply:
That's right!